bokeh/server/util.py

```diff
@@ -0,0 +1,204 @@
+                "values to restrict access explicitly", host)
+        if host == '*':
+            # do not append the :80 port suffix in that case: any port is
+            # accepted
+            hosts.append(host)
+            continue
+        parts = host.split(':')
+        if len(parts) == 1:
+            if parts[0] == "":
+                raise ValueError("Empty host value")
+            hosts.append(host+":80")
+        elif len(parts) == 2:
+            try:
+                int(parts[1])
+            except ValueError:
+                raise ValueError("Invalid port in host value: %s" % host)
+            if parts[0] == "":
+                raise ValueError("Empty host value")
+            hosts.append(host)
+        else:
+            raise ValueError("Invalid host value: %s" % host)
+    return hosts
+
+def match_host(host, pattern):
+    ''' Match a host string against a pattern
+
+    Args:
+        host (str)
+            A hostname to compare to the given pattern
+
+        pattern (str)
+            A string representing a hostname pattern, possibly including
+            wildcards for ip address octets or ports.
+
+    This function will return ``True`` if the hostname matches the pattern,
+    including any widcards. If the pattern contains a port, the host string
+    must also contain a matching port.
+
+    Returns:
+        bool
+
+    Examples:
+
+        >>> match_host('192.168.0.1:80', '192.168.0.1:80')
+        True
+        >>> match_host('192.168.0.1:80', '192.168.0.1')
+        True
+        >>> match_host('192.168.0.1:80', '192.168.0.1:8080')
+        False
+        >>> match_host('192.168.0.1', '192.168.0.2')
+        False
+        >>> match_host('192.168.0.1', '192.168.*.*')
+        True
+        >>> match_host('alice', 'alice')
+        True
+        >>> match_host('alice:80', 'alice')
+        True
+        >>> match_host('alice', 'bob')
+        False
+        >>> match_host('foo.example.com', 'foo.example.com.net')
+        False
+        >>> match_host('alice', '*')
+        True
+        >>> match_host('alice', '*:*')
+        True
+        >>> match_host('alice:80', '*')
+        True
+        >>> match_host('alice:80', '*:80')
+        True
+        >>> match_host('alice:8080', '*:80')
+        False
+
+    '''
+    if ':' in host:
+        host, host_port = host.rsplit(':', 1)
+    else:
+        host_port = None
+
+    if ':' in pattern:
+        pattern, pattern_port = pattern.rsplit(':', 1)
+        if pattern_port == '*':
+            pattern_port = None
+    else:
+        pattern_port = None
+
+    if pattern_port is not None and host_port != pattern_port:
+        return False
+
+    host = host.split('.')
+    pattern = pattern.split('.')
+
+    if len(pattern) > len(host):
+        return False
+
+    for h, p in zip(host, pattern):
+        if h == p or p == '*':
+            continue
+        else:
+            return False
+    return True
```

Comment:
These functions were scattered in other files. Collected them here and added tests.